### PR TITLE
Add Page filters

### PR DESCRIPTION
**Pull request by rynowak**
/cc @pranavkm

### CODE REVIEW

**Comment by rynowak** [comment]
Implementation to follow

**Comment by rynowak** [comment]
@pranavkm updated

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 1]
```diff
@@ -209,7 +209,7 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
 
                         filter.OnActionExecuted(actionExecutedContext);
 
-                        _diagnosticSource.BeforeOnActionExecuted(actionExecutedContext, filter);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: This was wrong :(

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Filters/IAsyncPageFilter.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        /// <summary>
+        /// Called asynchronously after the handler method has been selected, but before model binding occurs.
+        /// </summary>
+        /// <param name="context">The <see cref="PageHandlerSelectedContext"/>.</param>
+        /// <returns>A <see cref="Task"/> that on completion indicates the filter has executed.</returns>
+        Task OnPageHandlerSelectionAsync(PageHandlerSelectedContext context);
+
+        /// <summary>
+        /// Called asynchronously before the handler method, after model binding is complete.
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Filters/IAsyncPageFilter.cs]: > before the handler method is invoked, after

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 3]
```diff
@@ -190,39 +125,23 @@ private Task ExecutePageAsync()
 
                 _page = (Page)CacheEntry.PageFactory(_pageContext, _viewContext);
 
-                pageResult.Page = _page;
-                pageResult.ViewData = pageResult.ViewData ?? _pageContext.ViewData;
+                if (_actionDescriptor.ModelTypeInfo == _actionDescriptor.PageTypeInfo)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: `!HasPageModel`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 4]
```diff
@@ -298,16 +248,425 @@ private Task ExecutePageAsync()
                         break;
                     }
                 }
-                
-                result = await executor(instance, arguments);
+
+                var argsHack = new Dictionary<string, object>();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: o.o?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 5]
```diff
@@ -298,16 +248,425 @@ private Task ExecutePageAsync()
...
+                        _arguments = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+
+                        var task = BindArgumentsAsync();
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.PageNext;
+                            return task;
+                        }
+
+                        _cursor.Reset();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: Wouldn't you want to reset the cursor before the `if`?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs [hunk 6]
```diff
@@ -298,16 +248,425 @@ private Task ExecutePageAsync()
...
+                            if (_handlerExecutedContext == null)
+                            {
+                                _handlerExecutedContext = new PageHandlerExecutedContext(_pageContext, _filters, _handler, _instance)
+                                {
+                                    Result = _result,
+                                };
+                            }
+
+                            isCompleted = true;
+                            return TaskCache.CompletedTask;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: `Task.CompletedTask`.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs [hunk 7]
```diff
@@ -35,5 +36,29 @@ public PageActionDescriptor(PageActionDescriptor other)
...
         public string ViewEnginePath { get; set; }
+
+        /// <inheritdoc />
+        public override string DisplayName
+        {
+            get
+            {
+                if (base.DisplayName == null && ViewEnginePath != null)
+                {
+                    base.DisplayName = ViewEnginePath;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs]: Couldn't we just set this when constructing the descriptor?

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs [hunk 8]
```diff
@@ -22,17 +21,1066 @@
...
-using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Logging.Testing;
 using Moq;
 using Xunit;
 
 namespace Microsoft.AspNetCore.Mvc.RazorPages.Internal
 {
     public class PageActionInvokerTest : CommonResourceInvokerTest
     {
+        #region Diagnostics
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs]: Boo

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs [hunk 9]
```diff
@@ -22,17 +21,1066 @@
...
+        {
+            // Arrange
+            HandlerMethodDescriptor handler = null;
+
+            var filter1 = new Mock<IAsyncPageFilter>(MockBehavior.Strict);
+            filter1
+                .Setup(f => f.OnPageHandlerSelectionAsync(It.IsAny<PageHandlerSelectedContext>()))
+                .Callback<PageHandlerSelectedContext>(c =>
+                {
+                    handler = c.HandlerMethod = c.ActionDescriptor.HandlerMethods[1];
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs]: One assignment per line.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs [hunk 10]
```diff
@@ -22,17 +21,1066 @@
...
+            var message =
+                "If an IAsyncPageFilter provides a result value by setting the Result property of " +
+                "PageHandlerExecutingContext to a non-null value, then it cannot call the next filter by invoking " +
+                "PageHandlerExecutionDelegate.";
+
+            var invoker = CreateInvoker(pageFilter.Object);
+
+            // Act & Assert
+            await ExceptionAssert.ThrowsAsync<InvalidOperationException>(
+                async () => await invoker.InvokeAsync(),
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs]: `() => invoker.InvokeAsync()`

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs [hunk 11]
```diff
@@ -22,17 +21,1066 @@
...
+            var filter = new Mock<IPageFilter>(MockBehavior.Strict);
+            AllowSelector(filter);
+            filter.Setup(f => f.OnPageHandlerExecuting(It.IsAny<PageHandlerExecutingContext>())).Verifiable();
+            filter
+                .Setup(f => f.OnPageHandlerExecuted(It.IsAny<PageHandlerExecutedContext>()))
+                .Callback<PageHandlerExecutedContext>(c =>
+                {
+                    context = c;
+
+                    // Handle the exception so the test doesn't throw.
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs]: Can you Assert the exception instance?

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs [hunk 12]
```diff
@@ -22,17 +21,1066 @@
...
+                {
+                    resourceFilter.Object,
+                    pageFilter1.Object,
+                    pageFilter2.Object,
+                },
+                // The action won't run
+                exception: Exception);
+
+            // Act & Assert
+            await invoker.InvokeAsync();
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageActionInvokerTest.cs]: Can you verify the `ResourceFilter`. There's no way to tell if it got invoked.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: No, there are two cases where `!HasPageModel` is false. You can have handler methods on the page it self and still use `@model`.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: lol

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs]: yeah good catch

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs (hunk 7 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs]: I'm not sure how that would be better, just different.

**Pull request merged**